FWCore/Integration/test/TestInterProcessProd.cc

```diff
@@ -0,0 +1,286 @@
+          auto nlines = std::to_string(std::count(iConfig.begin(), iConfig.end(), '\n'));
+          auto result = fwrite(nlines.data(), sizeof(char), nlines.size(), pipe_);
+          assert(result = nlines.size());
+          result = fwrite(iConfig.data(), sizeof(char), iConfig.size(), pipe_);
+          assert(result == iConfig.size());
+          fflush(pipe_);
+        }
+      });
+    }
+
+    template <typename SERIAL>
+    auto doTransition(SERIAL& iDeserializer, edm::Transition iTrans, unsigned long long iTransitionID)
+        -> decltype(iDeserializer.deserialize()) {
+      decltype(iDeserializer.deserialize()) value;
+      if (not channel_.doTransition(
+              [&value, this]() {
+                value = deserializer_.deserialize();
+                std::cout << id_ << " from shared memory " << value.size() << std::endl;
+              },
+              iTrans,
+              iTransitionID)) {
+        std::cout << id_ << " FAILED waiting for external process" << std::endl;
+        externalFailed_ = true;
+        throw cms::Exception("ExternalFailed");
+      }
+      return value;
+    }
+    edmtest::ThingCollection produce(unsigned long long iTransitionID) {
+      return doTransition(deserializer_, edm::Transition::Event, iTransitionID);
+    }
+
+    edmtest::ThingCollection beginRunProduce(unsigned long long iTransitionID) {
+      return doTransition(br_deserializer_, edm::Transition::BeginRun, iTransitionID);
+    }
+
+    edmtest::ThingCollection endRunProduce(unsigned long long iTransitionID) {
+      if (not externalFailed_) {
+        return doTransition(er_deserializer_, edm::Transition::EndRun, iTransitionID);
+      }
+      return edmtest::ThingCollection();
+    }
+
+    edmtest::ThingCollection beginLumiProduce(unsigned long long iTransitionID) {
+      return doTransition(bl_deserializer_, edm::Transition::BeginLuminosityBlock, iTransitionID);
+    }
+
+    edmtest::ThingCollection endLumiProduce(unsigned long long iTransitionID) {
+      if (not externalFailed_) {
+        return doTransition(el_deserializer_, edm::Transition::EndLuminosityBlock, iTransitionID);
+      }
+      return edmtest::ThingCollection();
+    }
+
+    ~StreamCache() {
+      channel_.stopWorker();
+      pclose(pipe_);
+    }
+
+  private:
+    std::string unique_name(std::string iBase) {
+      auto pid = getpid();
+      iBase += std::to_string(pid);
+      iBase += "_";
+      iBase += std::to_string(id_);
+
+      return iBase;
+    }
+
+    int id_;
+    FILE* pipe_;
+    ControllerChannel channel_;
+    ReadBuffer readBuffer_;
+
+    using TCDeserializer = ROOTDeserializer<edmtest::ThingCollection, ReadBuffer>;
+    TCDeserializer deserializer_;
+    TCDeserializer br_deserializer_;
+    TCDeserializer er_deserializer_;
+    TCDeserializer bl_deserializer_;
+    TCDeserializer el_deserializer_;
+    bool externalFailed_ = false;
+  };
+
+  struct RunCache {
+    mutable edmtest::ThingCollection thingCollection_;
```

Comment:
Even if this is just a test, I'd suggest
```suggestion
    CMS_THREAD_SAFE mutable edmtest::ThingCollection thingCollection_;
```
with a comment explaining why the `mutable` is safe (I actually wonder why the static analyzer did not complain).